### ISSUE TITLE
[distributed] get-bucket-location hangs when N/2 servers are down

### ISSUE THREAD

**Issue by krishnasrinivas** [minio/minio#2685]
- start 3:3 minio distributed XL 
- copy a file into a bucket
- bring down 3 minio instances
- mc cat on the copied file hangs

This is the call that hangs:

```
REQUEST : GET path:/testbucket/?location=
Host: 127.0.0.1:9000
X-Amz-Date: 20160913T175226Z
Accept-Encoding: gzip
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/DEVELOPMENT.GOGET
Authorization: AWS4-HMAC-SHA256 Credential=SXO8VW2OFKKP2OG7AC85/20160913/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=4e1457da2e6092c8f4df2744d94b8241523fd16d597b74836a5b80be4b5048d8
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855


```

**Comment by harshavardhana** [comment]
> start 3:3 minio distributed XL
> copy a file into a bucket
> bring down 3 minio instances
> mc cat on the copied file hangs

@krishnasrinivas Locks are not granted with 3 servers going down this needs to be handled at dsync.  Dsync requires N/2+1 to be up to have a guarantee on locks.

**Comment by harshavardhana** [comment]
@krisis can you fill in ?

**Comment by krisis** [comment]
@krishnasrinivas @harshavardhana `dsync` blocks on `DRWMutex.[R]Lock()` until it acquires locks from quorum number of servers. It is expected to hang until the lock is acquired. 
@krishnasrinivas what is the expected behaviour in this situation?

**Comment by harshavardhana** [comment]
> @krishnasrinivas what is the expected behaviour in this situation?

The expected behavior that Reads continue since N/2 is only needed for read.

**Comment by krisis** [comment]
@harshavardhana The erasure code algorithm is capable of reconstructing data in the absence of up to N/2 disks. But to provide a consistent view of data, we need quorum (N/2+1) or more number of disks to be in agreement with the data that is being served. It makes sense to 'hang' the reads in favour of consistency. It isn't data loss, it is temporary data unavailability. if we replaced one of the disks, with a fresh one, we can read data without a 'hang'. Would it make sense to document this as the expected behaviour?

**Comment by harshavardhana** [comment]
> @harshavardhana The erasure code algorithm is capable of reconstructing data in the absence of up to N/2 disks. But to provide a consistent view of data, we need quorum (N/2+1) or more number of disks to be in agreement with the data that is being served. It makes sense to 'hang' the reads in favour of consistency. It isn't data loss, it is temporary data unavailability. if we replaced one of the disks, with a fresh one, we can read data without a 'hang'. Would it make sense to document this as the expected behaviour?

It would make sense if this is an agreed consensus, one problem here is that error in quorum is not propagated back to the API user since the locks themselves reject the request.

We should document this absolutely though.

**Comment by fwessels** [comment]
Let's reason through relaxing the quorum for RLocks to N/2 (while keeping (write) Locks at N/2+1):
- With N/2 there is a chance that (6 server example) one readlock is backed by 1, 2, and 3 and another readlock is backed by 4, 5, 6. This would be catastrophic for write locks but for read locks it is no problem since we allow multiple concurrent read locks anyhow). (Note that the situation is also quite unlikely since the locking process is 'aggressive' in the sense that it will try to get as much locks as possible).
- Continuing from the above, subsequent RLocks would be fine as long as still (any) combination of N/2 nodes is still up
- If a read lock relies only on N/2 nodes, an attempt for a write lock at N/2+1 will still fail (as desired)
- Continuing from above, if any one of the N/2 nodes that participated in the read lock crashes and restarts, then a quorum for write locks becomes available
- (with N/2+1 for read locks, this requires any two of the nodes that support a read lock to crash)
- When we have an (exclusive) write lock, still no read locks will be granted.
- If a write lock has just met quorum (N/2+1), and a single node out of these nodes goes down, all other nodes are up, then a read quorum becomes available (requires two nodes to go down for the current case).

So essentially changing this increases the likelyhood of concurrent locks in case of servers crashing or network down.
Note that given the 'aggressive' nature of the locking process, typically the tolerance for servers going down is much higher.

But still in the 3:3 example above, if 4 servers are still up then write locks will be granted and PUTs can happen -- if then another server goes down the write locks will be maintained and a read lock for the same resource can be granted provided that the (now 3) dead nodes are restarted.

**Comment by fwessels** [comment]
@harshavardhana So we want to relax the read quorom (from N/2+1) to N/2, correct?

(alternatively we could make this optional or configurable, ie starting minio in like a `--readonly` mode or so, but not sure if I am a fan of this...)

**Comment by harshavardhana** [comment]
> @harshavardhana So we want to relax the read quorom (from N/2+1) to N/2, correct?

Read quorum is already 'N/2',  only dsync choose N/2+1 for both `Lock` and `RLock`

```
        // Figure out read and write quorum based on number of storage disks.
        // READ and WRITE quorum is always set to (N/2) number of disks.
        xl.readQuorum = len(xl.storageDisks) / 2
        xl.writeQuorum = len(xl.storageDisks)/2 + 1
```

**Comment by fwessels** [comment]
Correct of course, I was referring to dsync's `RLock` behaviour specifically -- this only we want to relax to N/2, correct? (while keeping `Lock` unchanged)

**Comment by harshavardhana** [comment]
> Correct of course, I was referring to dsync's RLock behaviour specifically -- this only we want to relax to N/2, correct? (while keeping Lock unchanged)

Yeah since RLock is an okay operation but still we should think more about the cases where it might become problematic - do you remember why N/2+1 was chosen by default?

**Comment by fwessels** [comment]
Initially we only had Lock() and that needs (minimally) N/2 + 1 (and preferably more).

So for RLock() it was kept at the same level, but it is possible to differentiate between the two.

Let me do some coding/run some test cases and see if there are any stumbling blocks that we may have overlooked.

**Comment by GarimaKapoor** [comment]
@fwessels can you take a look at this?

**Comment by fwessels** [comment]
Yes, I will relax the read quorum (in dsync) to N/2

**Comment by fwessels** [comment]
As per https://github.com/minio/minio/pull/2874 the read lock quorum has been relaxed to N/2.

Testing this in a 2:2 cluster gives the desired behaviour:
- with 3 machines up we can copy data in
- with 2 machines up we can no longer copy data in (error reported after a short timeout):

```
$ mc cp 1MB.bin myminio/test
mc: <ERROR> Failed to copy ‘1MB.bin’. 503 Service Unavailable
mc: <ERROR> Session safely terminated. To resume session ‘mc session resume aKwEWmsk’
```
- with 2 machines up we can now happily do an `mc ls`:

```
$ mc ls myminio/test
[2016-10-07 09:16:02 UTC]  49MiB 50MB.bin
[2016-10-07 09:16:42 UTC] 4.9MiB 5MB.bin
```

and also copy data out

```
$ mc cp myminio/test/50MB.bin aap.out
.../50MB.bin:  48.83 MB / 48.83 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 84.93 MB/s 0
```

**Comment by harshavardhana** [comment]
Fixed by #2874
